README.md
# Farm_blast
Run blast+ or blastall in parallel on an LSF compute farm.

[![Build Status](https://travis-ci.org/sanger-pathogens/Farm_blast.svg?branch=master)](https://travis-ci.org/sanger-pathogens/Farm_blast)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/Farm_blast/blob/master/LICENSE)

## Contents
  * [Introduction](#introduction)
  * [Installation](#installation)
    * [Required dependencies](#required-dependencies)
    * [Optional dependencies](#optional-dependencies)
    * [Using pip](#using-pip)
    * [Running the tests](#running-the-tests)
  * [Usage](#usage)
  * [License](#license)
  * [Feedback/Issues](#feedbackissues)

## Introduction
Python3 module to run blast+ or blastall in parallel on an LSF compute farm.

## Installation
Farm_blast has the following dependencies:

### Required dependencies
  * [blast+](http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
### Optional dependencies
  * blastall

Details for installing Farm_blast are provided below. If you encounter an issue when installing Farm_blast please contact your local system administrator. If you encounter a bug please log it [here](https://github.com/sanger-pathogens/Farm_blast/issues) or email us at path-help@sanger.ac.uk.

### Using pip
`pip3 install farm_blast`

### Running the tests
The test can be run from the top level directory:   
`python3 setup.py test`

## Usage
Compare a query and a reference using blast+ blastn:

`farm_blast reference.fasta query.fasta`

Compare a query and a reference using blastall blastn:

`farm_blast --blastall reference.fasta query.fasta`

Run blast+ megablast:

`farm_blast --blast_type megablast reference.fasta query.fasta`

Run blastall tblastx:

`farm_blast --blast_type tblastx --blastall reference.fasta query.fasta`

Set the e-value and word length and do not filter the query sequence:

`farm_blast --no_filter -e 0.1 -W 30 reference.fasta query.fasta`

To get all the options, use `--help`:

`farm_blast --help`

## License
Farm_blast is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/Farm_blast/blob/master/LICENSE).

## Feedback/Issues
Please report any issues to the [issues page](https://github.com/sanger-pathogens/Farm_blast/issues) or email path-help@sanger.ac.uk.

CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/sanger-pathogens/Farm_blast/tree/HEAD)

[Full Changelog](https://github.com/sanger-pathogens/Farm_blast/compare/v0.1.2...HEAD)

**Merged pull requests:**

- Update readme [\#19](https://github.com/sanger-pathogens/Farm_blast/pull/19) ([ssjunnebo](https://github.com/ssjunnebo))
- 574328 add travis [\#17](https://github.com/sanger-pathogens/Farm_blast/pull/17) ([ssjunnebo](https://github.com/ssjunnebo))
- update AUTHORS and LICENSE [\#16](https://github.com/sanger-pathogens/Farm_blast/pull/16) ([ssjunnebo](https://github.com/ssjunnebo))

## [v0.1.2](https://github.com/sanger-pathogens/Farm_blast/tree/v0.1.2) (2015-02-02)
**Merged pull requests:**

- Change to pyfastaq; add to pypi [\#15](https://github.com/sanger-pathogens/Farm_blast/pull/15) ([martinghunt](https://github.com/martinghunt))
- version number inc for farm blast [\#14](https://github.com/sanger-pathogens/Farm_blast/pull/14) ([andrewjpage](https://github.com/andrewjpage))
- update fastaq underscore [\#13](https://github.com/sanger-pathogens/Farm_blast/pull/13) ([andrewjpage](https://github.com/andrewjpage))
- Added AUTHORS file [\#12](https://github.com/sanger-pathogens/Farm_blast/pull/12) ([aslett1](https://github.com/aslett1))
- Fix blast types [\#11](https://github.com/sanger-pathogens/Farm_blast/pull/11) ([martinghunt](https://github.com/martinghunt))
- Require fastaq \>= 1.2 [\#10](https://github.com/sanger-pathogens/Farm_blast/pull/10) ([martinghunt](https://github.com/martinghunt))
- Add --act option to make union files [\#9](https://github.com/sanger-pathogens/Farm_blast/pull/9) ([martinghunt](https://github.com/martinghunt))
- Ask for 2 threads in final combine job [\#8](https://github.com/sanger-pathogens/Farm_blast/pull/8) ([martinghunt](https://github.com/martinghunt))
- Add blast program version and type to output directory name [\#7](https://github.com/sanger-pathogens/Farm_blast/pull/7) ([martinghunt](https://github.com/martinghunt))
- Fix bugs with blastall megablast adding header lines to its output, also... [\#6](https://github.com/sanger-pathogens/Farm_blast/pull/6) ([martinghunt](https://github.com/martinghunt))
- Fix bug where pipeline died if a split query chunk was all Ns [\#5](https://github.com/sanger-pathogens/Farm_blast/pull/5) ([martinghunt](https://github.com/martinghunt))
- Support other file formats [\#4](https://github.com/sanger-pathogens/Farm_blast/pull/4) ([martinghunt](https://github.com/martinghunt))
- Use basename instead of full path of input files to make output director... [\#3](https://github.com/sanger-pathogens/Farm_blast/pull/3) ([martinghunt](https://github.com/martinghunt))
- Fix typos in README [\#2](https://github.com/sanger-pathogens/Farm_blast/pull/2) ([martinghunt](https://github.com/martinghunt))
- Initial add of files [\#1](https://github.com/sanger-pathogens/Farm_blast/pull/1) ([martinghunt](https://github.com/martinghunt))




farm_blast/tests/blast_test.py
#!/usr/bin/env python3

import sys
import os
import filecmp
import unittest
from farm_blast import blast

modules_dir = os.path.dirname(os.path.abspath(blast.__file__))
data_dir = os.path.join(modules_dir, 'tests', 'data')

class TestBlast(unittest.TestCase):
    def test_check_blast_type(self):
        '''Check dies if blast_type not recognised'''
        with self.assertRaises(blast.Error):
            b = blast.Blast('query.fasta', 'ref.fasta', blast_type='oops')
        

    def test_blast_db_exists(self):
        '''Test detection or not of blast database'''
        query_file = 'tmp.test_blast_db_exists.fa'
        b_blastn = blast.Blast(query_file, 'ref.fasta')
        b_blastp = blast.Blast(query_file, 'ref.fasta', blast_type='blastp')

        nuc_suffixes = ['nin', 'nhr', 'nsq']
        nuc_suffixes2 = ['00.' + x for x in nuc_suffixes]
        pro_suffixes = ['pin', 'phr', 'psq']
        pro_suffixes2 = ['00.' + x for x in pro_suffixes]

        open(query_file, 'w').close()
        self.assertFalse(b_blastn.blast_db_exists())
        self.assertFalse(b_blastp.blast_db_exists())

        tuples = [
            (nuc_suffixes, b_blastn, b_blastp),
            (nuc_suffixes2, b_blastn, b_blastp),
            (pro_suffixes, b_blastp, b_blastn),
            (pro_suffixes2, b_blastp, b_blastn)
        ]
                  
        for suffixes, blast1, blast2 in tuples:
            print(suffixes, blast1.blast_type, blast2.blast_type)
            for suff in suffixes:
                open(query_file + '.' + suff, 'w').close()

            self.assertTrue(blast1.blast_db_exists())
            self.assertFalse(blast2.blast_db_exists())
     
            for suff in suffixes:
                missing_file = query_file + '.' + suff
                os.unlink(missing_file)
                self.assertFalse(blast1.blast_db_exists())
                open(missing_file, 'w').close()
                
            for suff in suffixes:
                os.unlink(query_file + '.' + suff)
            
        os.unlink(query_file)


    def test_format_database_command(self):
        '''Test command to format database made correctly for blast+ and blastall'''
        b_all_nuc = blast.Blast('ref.fasta', 'qry.fasta', blastall=True)
        b_all_pro = blast.Blast('ref.fasta', 'qry.fasta', blast_type='blastp', blastall=True)
        b_plus_nuc = blast.Blast('ref.fasta', 'qry.fasta')
        b_plus_pro = blast.Blast('ref.fasta', 'qry.fasta', blast_type='blastp')

        blastall_nuc = 'formatdb -p F -i ref.fasta'
        blastall_pro = 'formatdb -p T -i ref.fasta'
        blastplus_nuc = 'makeblastdb -dbtype nucl -in ref.fasta'
        blastplus_pro = 'makeblastdb -dbtype prot -in ref.fasta'
        
        self.assertEqual(b_all_nuc.format_database_command(), blastall_nuc)
        self.assertEqual(b_all_pro.format_database_command(), blastall_pro)
        self.assertEqual(b_plus_nuc.format_database_command(), blastplus_nuc)
        self.assertEqual(b_plus_pro.format_database_command(), blastplus_pro)

        query_file = 'tmp.test_blast_db_exists.fa'
        b_blastn = blast.Blast(query_file, 'ref.fasta')
        files = [query_file + '.' + x for x in ['nin', 'nhr', 'nsq']]
        for f in files:
            open(f, 'w').close()
        self.assertEqual(b_blastn.format_database_command(), None)
        for f in files:
            os.unlink(f)


    def test_make_options_string(self):
        '''Test options set correctly'''
        test_objs = [
            blast.Blast('ref', 'qry'),
            blast.Blast('ref', 'qry', evalue=0.1),
            blast.Blast('ref', 'qry', word_size=42),
            blast.Blast('ref', 'qry', no_filter=True),
            blast.Blast('ref', 'qry', extra_options='-x 1'),
            blast.Blast('ref', 'qry', blast_type='blastp'),
            blast.Blast('ref', 'qry', blastall=True),
            blast.Blast('ref', 'qry', blastall=True, evalue=0.1),
            blast.Blast('ref', 'qry', blastall=True, word_size=42),
            blast.Blast('ref', 'qry', blastall=True, no_filter=True),
        ]

        correct = [
            '-outfmt 6',
            '-outfmt 6 -evalue 0.1',
            '-outfmt 6 -word_size 42',
            '-outfmt 6 -dust no',
            '-outfmt 6 -x 1',
            '-outfmt 6 -seg yes',
            '-m 8',
            '-m 8 -e 0.1',
            '-m 8 -W 42',
            '-m 8 -F F'
        ]

        for i in range(len(correct)):
            self.assertEqual(correct[i], test_objs[i]._make_options_string())


    def test_make_io_string(self):
        '''Test input/output files string'''
        test_objs = [
            blast.Blast('ref', 'qry'),
            blast.Blast('ref', 'qry', blastall=True),
        ]

        correct = [
            '-db ref -query qry -out blast.out',
            '-d ref -i qry -o blast.out',
        ]
 
        for i in range(len(correct)):
            self.assertEqual(correct[i], test_objs[i]._make_io_string())


    def test_make_blast_type_string(self):
        '''Test blast type string'''
        test_objs = [
            (blast.Blast('ref', 'qry'), 'blastn -task blastn'),
            (blast.Blast('ref', 'qry', blast_type='blastn'), 'blastn -task blastn'),
            (blast.Blast('ref', 'qry', blast_type='blastn-short'), 'blastn -task blastn-short'),
            (blast.Blast('ref', 'qry', blast_type='dc-megablast'), 'blastn -task dc-megablast'),
            (blast.Blast('ref', 'qry', blast_type='megablast'), 'blastn -task megablast'),
            (blast.Blast('ref', 'qry', blast_type='rmblastn'), 'blastn -task rmblastn'),
            (blast.Blast('ref', 'qry', blast_type='blastx'), 'blastx'),
            (blast.Blast('ref', 'qry', blast_type='blastp'), 'blastp -task blastp'),
            (blast.Blast('ref', 'qry', blast_type='blastp-short'), 'blastp -task blastp-short'),
            (blast.Blast('ref', 'qry', blast_type='deltablast'), 'blastp -task deltablast'),
            (blast.Blast('ref', 'qry', blast_type='tblastn'), 'tblastn'),
            (blast.Blast('ref', 'qry', blast_type='tblastx'), 'tblastx'),
            (blast.Blast('ref', 'qry', blastall=True), 'blastall -p blastn'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='blastn'), 'blastall -p blastn'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='blastx'), 'blastall -p blastx'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='blastp'), 'blastall -p blastp'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='tblastn'), 'blastall -p tblastn'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='tblastx'), 'blastall -p tblastx'),
            (blast.Blast('ref', 'qry', blastall=True, blast_type='megablast'), 'blastall -p blastn -n T')
        ]
        
        for t in test_objs:
            self.assertEqual(t[1], t[0]._make_blast_type_string())

        with self.assertRaises(blast.Error):
            b = blast.Blast('ref', 'qry', blast_type='oops')

    def test_get_run_command(self):
        '''Test command to run blast made OK'''
        b = blast.Blast('qry.fasta', 'ref.fasta', evalue=0.1)
        expected = ' '.join([
            b._make_blast_type_string(),
            b._make_io_string(),
            b._make_options_string()
        ])

        self.assertEqual(expected, b.get_run_command())

if __name__ == '__main__':
    unittest.main()


farm_blast/tests/utils_test.py
#!/usr/bin/env python3

import sys
import os
import filecmp
import unittest
from farm_blast import utils


modules_dir = os.path.dirname(os.path.abspath(utils.__file__))
data_dir = os.path.join(modules_dir, 'tests', 'data')

class TestUtils(unittest.TestCase):
    def test_offset_coords_file_to_dict(self):
        '''Test coords offsets load into dict from file'''
        correct_coords = {
            'seq1': ('ref1', 0),
            'seq2': ('ref1', 10),
            'seq3': ('ref2', 0),
            'seq4': ('ref2', 42)
        }

        d = utils.offset_coords_file_to_dict(os.path.join(data_dir, 'utils_test_coords_offset.tsv'))
        self.assertDictEqual(d, correct_coords)


    def test_fix_blast_coords(self):
        '''Test blast coords fixed correctly'''
        blast_file = os.path.join(data_dir, 'utils_test_fix_blast_coords.blast')
        coords_file = os.path.join(data_dir, 'utils_test_coords_offset.tsv')
        outfile = 'tmp.fix_blast_coords.out'
        utils.fix_blast_coords(blast_file, coords_file, outfile)
        self.assertTrue(filecmp.cmp(os.path.join(data_dir, 'utils_test_fix_blast_coords.blast.fixed'), outfile))
        os.unlink(outfile)

if __name__ == '__main__':
    unittest.main()


farm_blast/tests/pipeline_test.py
#!/usr/bin/env python3

import sys
import os
import filecmp
import subprocess
import shutil
import unittest
from farm_blast import pipeline

modules_dir = os.path.dirname(os.path.abspath(pipeline.__file__))
data_dir = os.path.join(modules_dir, 'tests', 'data')

class TestPipeline(unittest.TestCase):
    def setUp(self):
        self.ref = os.path.join(data_dir, 'pipeline_test.ref.fa')
        self.qry = os.path.join(data_dir, 'pipeline_test.qry.fa')
        options = pipeline.get_opts(args=[
            '--no_bsub',
            '--split_bases_tolerance', '1',
            '--split_bases', '100',
            '--outdir', 'tmp.Farm_blast_test',
            '--bsub_name_prefix', 'name',
            '--test',
            self.ref,
            self.qry])
        self.p = pipeline.Pipeline(options, os.path.abspath('scripts/farm_blast'))


    def test_make_setup_script_ref_already_indexed(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        ref = os.path.join(data_dir, 'pipeline_test.ref.indexed.fa')
        options = pipeline.get_opts(args=[
            '--no_bsub',
            '--split_bases_tolerance', '1',
            '--split_bases', '100',
            '--outdir', 'tmp.Farm_blast_test',
            '--bsub_name_prefix', 'name',
            ref,
            self.qry])
        self.p = pipeline.Pipeline(options, os.path.abspath('scripts/farm_blast'))
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.qry), '- |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_script_convert_ref(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.ref), 'reference.fa', file=f)
        print('makeblastdb -dbtype nucl -in reference.fa', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.qry), '- |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_script_with_union_for_act(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        self.p.union_for_act = True
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq merge', os.path.abspath(self.ref), '- |',
              'fastaq to_fasta -s - reference.fa', file=f)
        print('makeblastdb -dbtype nucl -in reference.fa', file=f)
        print('fastaq merge', os.path.abspath(self.qry), '- |',
              'fastaq to_fasta -s - - |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_job(self):
        self.p._make_setup_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-R "select[mem>1000] rusage[mem=1000]" -M1000 -o 01.setup.sh.o -e 01.setup.sh.e -J name.setup bash 01.setup.sh'''
        self.assertTrue(str(self.p.setup_job).endswith(expected))


    def test_make_start_array_job(self):
        self.p._make_start_array_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-o 02.run_array.sh.o -e 02.run_array.sh.e -J name.start_array bash 02.run_array.sh'''
        self.assertTrue(str(self.p.start_array_job).endswith(expected))


    def test_make_array_job(self):
        self.p._make_array_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-R "select[mem>500] rusage[mem=500]" -M500 -o tmp.array.o.%I -e tmp.array.e.%I -J "name.array[1-$n]%100" blastn -task blastn -db ''' + self.ref + ''' -query query.split.\$LSB_JOBINDEX -out tmp.array.out.\$LSB_JOBINDEX -outfmt 6'''
        self.assertTrue(str(self.p.array_job).endswith(expected))


    def test_make_start_array_script(self):
        expected_script = os.path.join(data_dir, 'pipeline_test.02.array.sh')
        test_script = 'tmp.make_start_array_script_test'
        self.p.array_job = 'array_job'
        self.p._make_start_array_script(script_name=test_script)
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(test_script)


    def test_make_combine_job(self):
        self.p._make_combine_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-n 2 -R "span[hosts=1] select[mem>500] rusage[mem=500]" -M500 -o 03.combine.sh.o -e 03.combine.sh.e -J name.combine bash 03.combine.sh'''
        self.assertTrue(str(self.p.combine_job).endswith(expected))


    def test_make_combine_script(self):
        expected_script = os.path.join(data_dir, 'pipeline_test.03.combine.sh')
        test_script = 'tmp.make_combine_array_script_test'
        self.p._make_combine_script(script_name=test_script)
        expected = [
            'cat tmp.array.e.* > 02.array.e',
            'cat tmp.array.o.* > 02.array.o',
            'cat tmp.array.out.* | gzip -9 -c > blast.out.tmp.gz',
            'PYTHONPATH=',
            '--fix_coords_in_blast_output x x',
            'rm tmp.array.* query.split.* blast.out.tmp.gz 02.array.id 03.combine.sh.id',
            'touch FINISHED'
        ]

        try:
            f = open(test_script)
        except:
            sys.exit('Error opening file', test_script)

        got = [x.rstrip() for x in f.readlines()]
        f.close()

        # the script has the absolute path to farm_blast in it. We don't know
        # where that will be, so ignore that part
        self.assertEqual(len(got), len(expected))
        self.assertTrue(got[3].startswith('PYTHONPATH'))
        self.assertTrue('farm_blast' in got[4])
        got[4] = got[4].split(None,1)[1]
        self.assertListEqual(got[:2], expected[:2])
        self.assertListEqual(got[5:], expected[5:])
        os.unlink(test_script)

    def test_pipeline(self):
        self.p.run()
        expected = os.path.join(data_dir, 'pipeline_test.blast.out')
        subprocess.call('gunzip ' + os.path.join(self.p.outdir, 'blast.out.gz'), shell=True)
        expected_list = []
        with open(expected,'r') as f:
            for line in f:
                row = line.split()
                expected_list.append([row[0:2],row[3:9],row[11]])
        gotten = os.path.join(self.p.outdir, 'blast.out')
        gotten_list = []
        with open(gotten, 'r') as f:
            for line in f:
                row = line.split()
                gotten_list.append([row[0:2],row[3:9],row[11]])
        self.assertEqual(expected_list, gotten_list)

    def tearDown(self):
        if os.path.exists(self.p.outdir):
            shutil.rmtree(self.p.outdir)


farm_blast/pipeline.py
import os
import sys
import argparse
import time
import glob
from farmpy import lsf
from farm_blast import blast, utils

class Error (Exception): pass

parser = argparse.ArgumentParser(
    description = 'Run BLAST in parallel on the farm',
    usage = '%(prog)s [options] <reference> <query>')

parser.add_argument('--no_bsub', action='store_true', help=argparse.SUPPRESS)
parser.add_argument('--fix_coords_in_blast_output', action='store_true', help=argparse.SUPPRESS)
parser.add_argument('--split_bases_tolerance', type=int, default=1000, help=argparse.SUPPRESS)
parser.add_argument('--test', action='store_true', help=argparse.SUPPRESS)

common_blast_group = parser.add_argument_group('Common BLAST options')
common_blast_group.add_argument('--blastall', action='store_true', help='Use blastall instead of the default blast+')
common_blast_group.add_argument('-p', '--blast_type', help='Type of blast to run [%(default)s]', choices=sorted(list(blast.ref_not_protein_types)) + sorted(list(blast.ref_protein_types)), default='blastn')
common_blast_group.add_argument('-e', '--evalue', help='Set the evalue cutoff')
common_blast_group.add_argument('-W', '--word_size', help='Set the word size')
common_blast_group.add_argument('--no_filter', action='store_true', help='Do not filter query sequence (equivalent to -F F in blastall, -dust no in blast+). By default, the query will be filtered')

bsub_group = parser.add_argument_group('Bsub options')
bsub_group.add_argument('-q', '--bsub_queue', help='Queue in which all jobs are run [%(default)s]', default = 'normal', metavar='Queue_name')
bsub_group.add_argument('--blast_mem', type=float, help='Memory limit in GB for the farm jobs that run BLAST. Default is 0.5, except set to 5 if blastall tblastx is used. Defaults doubled if --no_filter used', metavar='FLOAT', default=None)
bsub_group.add_argument('--bsub_name_prefix', help='Set the prefix of the names of the bsub jobs', default=None)


advanced_opts_group = parser.add_argument_group('Advanced options')
advanced_opts_group.add_argument('--act', action='store_true', help='Make ACT-friendly blast file, by concatenating all reference sequences together and all query sequences together before blasting.')
advanced_opts_group.add_argument('--blast_options', help='Put any extra options to the blast call (i.e. blastall, blastn, blastx ...etc) in quotes. e.g. --blast_options "-r 2". Whatever you put in here is NOT sanity checked.', default = '', metavar='"options in quotes"')
advanced_opts_group.add_argument('--debug', action='store_true', help='Just make scripts etc but do not run anything')
advanced_opts_group.add_argument('--outdir', help='Name of output directory (must not exist already)', metavar='output directory', default=None)
advanced_opts_group.add_argument('--split_bases', type=int, help='Number of bases in each split file of query. Default is 500000, except set to 200000 if blastall tblastx is used', metavar='INT', default=None)

parser.add_argument('reference', help='Name of reference file. Does not need to be indexed already. If not indexed, can be any format from FASTA, FASTQ, GFF3, EMBL, Phylip, GBK', metavar='reference')
parser.add_argument('query', help='Name of query file. Can be any format from FASTA, FASTQ, GFF3, EMBL, Phylip, GBK', metavar='query')

def get_opts(args=None):
    return parser.parse_args(args=args)


class Pipeline:
    def __init__(self, options, farm_blast_script):
        if options.outdir is None:
            if options.blastall:
                version = 'blastall'
            else:
                version = 'blast_plus'

            options.outdir = '.'.join(['Farm_blast', os.path.basename(options.reference), os.path.basename(options.query), version, options.blast_type, 'out'])

        self.outdir = os.path.abspath(options.outdir)
        self.reference = os.path.abspath(options.reference)
        self.query = os.path.abspath(options.query)
        self.bsub_queue = options.bsub_queue
        self.farm_blast_script = farm_blast_script
        self.test = options.test
        self.union_for_act = options.act

        self.blast = blast.Blast(
            self.reference,
            'query.split.INDEX',
            outfile='tmp.array.out.INDEX',
            blastall=options.blastall,
            blast_type=options.blast_type,
            evalue=options.evalue,
            word_size=options.word_size,
            no_filter=options.no_filter,
            extra_options=options.blast_options
        )

        self.setup_script = '01.setup.sh'
        self.start_array_script = '02.run_array.sh'
        self.combine_script = '03.combine.sh'
        if options.bsub_name_prefix is None:
            self.bsub_name_prefix = 'farm_blast:' + self.outdir
        else:
            self.bsub_name_prefix = options.bsub_name_prefix

        if options.no_bsub:
            self.no_bsub = True
            self.memory_units = 'MB'
        else:
            self.no_bsub = False
            self.memory_units = None

        self.debug = options.debug
        self.split_bases_tolerance = options.split_bases_tolerance

        self.files_to_delete = [
            'tmp.array.*',
            'query.split.*',
            'blast.out.tmp.gz',
            '02.array.id',
            '03.combine.sh.id',
        ]

        if not options.blast_mem:
            if self.blast.blastall and self.blast.blast_type == 'tblastx':
                self.array_mem = 5
            else:
                self.array_mem = 0.5

            if self.blast.no_filter:
                self.array_mem *= 2
        else:
            self.array_mem = options.blast_mem

        if not options.split_bases:
            if self.blast.blastall and self.blast.blast_type == 'tblastx':
                self.split_bases = 200000
            else:
                self.split_bases = 500000
        else:
            self.split_bases = options.split_bases


    def _make_setup_script(self, script_name=None):
        if script_name is None:
            script_name = self.setup_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error opening setup script "' + script_name + '" for writing')

        print('set -e', file=f)
        if not self.blast.blast_db_exists() or self.union_for_act:
            if self.union_for_act:
                print('fastaq merge', self.reference, '- |',
                      'fastaq to_fasta -s - reference.fa', file=f)
            else:
                print('fastaq to_fasta -s', self.reference, 'reference.fa', file=f)
            self.reference = 'reference.fa'
            self.blast.reference = self.reference
            print(self.blast.format_database_command(), file=f)
            self.files_to_delete.append('reference.*')

        # blast strips off everything after the first whitespace, so do this
        # before chunking so names stay consistent with query fasta and in blast output
        if self.union_for_act:
            print('fastaq merge', self.query, '- |',
                  'fastaq to_fasta -s - - |', end=' ', file=f)
        else:
            print('fastaq to_fasta -s', self.query, '- |', end=' ', file=f)

        print('fastaq chunker --skip_all_Ns', '-', 'query.split', self.split_bases, self.split_bases_tolerance, file=f)
        f.close()


    def _make_setup_job(self):
        self.setup_job = lsf.Job(
            self.setup_script + '.o',
            self.setup_script + '.e',
            self.bsub_name_prefix + '.setup',
            self.bsub_queue,
            1,
            'bash ' + self.setup_script,
            memory_units=self.memory_units,
        )


    def _make_array_job(self):
        self.array_job = lsf.Job(
            'tmp.array.o',
            'tmp.array.e',
            self.bsub_name_prefix + '.array',
            self.bsub_queue,
            self.array_mem,
            self.blast.get_run_command(),
            array_start=1,
            array_end=r'''$n''',
            memory_units=self.memory_units,
            max_array_size=100
        )


    def _make_start_array_job(self):
        self.start_array_job = lsf.Job(
            self.start_array_script + '.o',
            self.start_array_script + '.e',
            self.bsub_name_prefix + '.start_array',
            'small',
            0.1,
            'bash ' + self.start_array_script,
            no_resources=True  # we have to do this otherwise bmod fails! LSF bug?
        )


    def _make_start_array_script(self, script_name=None):
        if script_name is None:
            script_name = self.start_array_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error writing script "' + script_name + '"')
        print('set -e', file=f)
        print(r'''n=`ls query.split.* | grep -v coords | wc -l`''', file=f)
        print(str(self.array_job) + r''' |  awk '{print substr($2,2,length($2)-2)}' > 02.array.id''', file=f)
        print(r'''array_id=`cat 02.array.id`
combine_id=`cat ''' + self.combine_script + r'''.id`
bmod -w "done($array_id)" $combine_id''', file=f)
        f.close()


    def _make_combine_job(self):
        self.combine_job = lsf.Job(
            self.combine_script + '.o',
            self.combine_script + '.e',
            self.bsub_name_prefix + '.combine',
            self.bsub_queue,
            0.5,
            'bash ' + self.combine_script,
            memory_units=self.memory_units,
            threads=2
        )


    def _make_combine_script(self, script_name=None):
        if script_name is None:
            script_name = self.combine_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error writing script "' + script_name + '"')

        if not self.no_bsub:
            print('set -e', file=f)

        print(r'''cat tmp.array.e.* > 02.array.e
cat tmp.array.o.* > 02.array.o
cat tmp.array.out.* | gzip -9 -c > blast.out.tmp.gz''', file=f)
        if self.test:
            p = os.path.dirname(self.farm_blast_script)
            p = os.path.join(p, os.pardir)
            p = os.path.normpath(p)
            print('PYTHONPATH=' + p + ':$PYTHONPATH', file=f)
        if self.test:
            print(self.farm_blast_script, '--test --fix_coords_in_blast_output x x', file=f)
        else:
            print(self.farm_blast_script, '--fix_coords_in_blast_output x x', file=f)
        print('rm', ' '.join(self.files_to_delete), file=f)
        print('touch FINISHED', file=f)
        f.close()


    def run(self):
        '''Runs the whole blast pipeline'''
        # Here's the fun part: the first job splits the query file into
        # chunks. Don't know the number of chunks until that job has finished.
        # Therefore don't know the size of the job array until the chunking
        # job has finished.
        #
        # Here's what's going to happen:
        # 1. Submit chunking job.
        # 2. Submit job to submit an array, to run when Job1 finishes.
        #      - this is NOT the array itself!
        # 3. Submit job that combines the results of the job array and tidies
        # up files. It is initially set to run when the job that starts
        # the array finishes.
        # 4. When the job that starts the array actually runs, it does this:
        #       - figures out the size of the job array and submits it
        #       - changes the dependency of the last combine job, so that
        #         the combien job runs when the array has finished
        # 5. When the array finishes, Job3 will then run.

        try:
            os.mkdir(self.outdir)
        except:
            print('Error making output directory', self.outdir, file=sys.stderr)
            sys.exit(1)

        original_dir = os.getcwd()
        os.chdir(self.outdir)
        self._make_setup_script()
        self._make_setup_job()
        self._make_array_job()
        self._make_start_array_script()
        self._make_start_array_job()
        self._make_combine_script()
        self._make_combine_job()

        if self.debug:
            sys.exit()

        if self.no_bsub:
            self.setup_job.run_not_bsubbed()
            # get size of job array
            files_count = len(glob.glob('query.split.*')) - 1
            self.array_job.array_end = files_count
            print(self.array_job)
            print(self.array_job.array_start)
            print(self.array_job.array_end)
            print(self.array_job._make_command_string().replace('\$LSB_JOBINDEX', '$LSB_JOBINDEX'))
            self.array_job.run_not_bsubbed()

            # a little hack here to make the farm_blast script run
            this_script = os.path.realpath(__file__)
            this_script_dir = os.path.dirname(this_script)
            module_root_dir = os.path.join(this_script_dir, '..')
            module_root_dir = os.path.normpath(module_root_dir)
            os.environ["PATH"] = os.path.join(module_root_dir, 'scripts:') + os.environ["PATH"]
            os.environ["PYTHONPATH"] = module_root_dir + ':' + os.environ["PYTHONPATH"]
            self.combine_job.run_not_bsubbed()
        else:
            self.setup_job.run()
            time.sleep(1)
            self.start_array_job.add_dependency(self.setup_job.job_id)
            self.start_array_job.run()
            time.sleep(1)
            self.combine_job.add_dependency(self.start_array_job.job_id)
            self.combine_job.run()
            time.sleep(1)
            try:
                f = open(self.combine_script + '.id', 'w')
            except:
                raise Error('Error opening file "' + self.combine_script + '.id' + '" for writing')
            print(self.combine_job.job_id, file=f)
            f.close()
            print('Jobs submitted to the farm.')
            print('Final job id is', self.combine_job.job_id)
            print('\nPipeline finished OK when this file is written:\n   ', os.path.join(self.outdir, 'FINISHED'))
            print('\nFinal file will be called:\n   ', os.path.join(self.outdir, 'blast.out.gz'))

        os.chdir(original_dir)
